### PR TITLE
Add jacobian and hessian

### PR DESCRIPTION
**Pull request by HydrogenSulfate**
### PR types

New features

### PR changes

APIs

### Description

Pcard-66961

Add 2 high-order differential API: `paddle.autograd.jacobian` 与 `paddle.autograd.hessian`，and related unitest file.

Move paddle.incubate.autograd.Jacobian/Hessian to paddle.incubate.autograd.functional.Jacobian/Hessian

### CODE REVIEW

**Comment by cxxly** [comment]
rename autodiff to autograd to keep  consistent with the convention of paddle.

File: python/paddle/autograd/autodiff.py [hunk 1]
```diff
@@ -0,0 +1,748 @@
...
+    Once the Jacobian ``J`` is constructed, you can use a multidimensional index
+    to retrieve the submatrix of ``J``, as same as slicing a Tensor. The
+    submatrix is lazily evaluated along row axis, and will be cached once
+    evaluated.
+
+    For examples, supposing ``is_batched=True``, and ys is a single Tensor with shape of
+    ``[B, D]``, xs is a single Tensor with shape of ``[B, C]``, and result jacobian
+    matrix is with shape of ``[B, D, C]``, then you can retrieve the submatrix by
+    following methods:
+
```
Comment (cxxly) [inline python/paddle/autograd/autodiff.py]: fix doc description.

File: python/paddle/autograd/autodiff.py [hunk 2]
```diff
@@ -0,0 +1,748 @@
...
+    def __ge__(self, other):
+        lhs = self._evaluate_all()
+        rhs = other._evaluate_all() if isinstance(other, Jacobian) else other
+        return lhs >= rhs
+
+
+class Hessian(Jacobian):
+    """Hessian class, same as Jacobian except classname"""
+
```
Comment (cxxly) [inline python/paddle/autograd/autodiff.py]: add comment to explain why need this empty class, or just remove it.

File: python/paddle/autograd/autodiff.py [hunk 3]
```diff
@@ -0,0 +1,748 @@
...
+        ys (Union[paddle.Tensor, Tuple[paddle.Tensor, ...]]): Output or list of outputs derived from xs.
+        xs (Union[paddle.Tensor, Tuple[paddle.Tensor, ...]]): Input or list of inputs.
+        batch_axis (Optional[int], optional): Index of batch axis. Defaults to None.
+
+    Returns:
+
+        Union[List[List[Jacobian]], List[Jacobian], Jacobian]: Jacobian(s) of ys
+            deriveted from xs.
+
```
Comment (cxxly) [inline python/paddle/autograd/autodiff.py]: Tuple

File: python/paddle/autograd/autodiff.py [hunk 4]
```diff
@@ -0,0 +1,748 @@
...
+        raise ValueError(
+            f"batch_axis should be None or 0, but got {batch_axis}."
+        )
+
+    # TODO(HydrogenSulfate): support batch_axis > 0
+    is_batched = batch_axis is not None
+    if isinstance(ys, Sequence) and isinstance(xs, Sequence):
+        _jacobian = [
+            [Jacobian(_ys, _xs, is_batched) for _xs in xs] for _ys in ys
+        ]
```
Comment (cxxly) [inline python/paddle/autograd/autodiff.py]: use tuple of tuple

File: python/paddle/autograd/autodiff.py [hunk 5]
```diff
@@ -0,0 +1,748 @@
...
+    return _jacobian
+
+
+def hessian(
+    ys: paddle.Tensor,
+    xs: Union[paddle.Tensor, Tuple[paddle.Tensor, ...]],
+    batch_axis: Optional[int] = None,
+) -> Union[List[List[Hessian]], Hessian]:
+    r"""Computes the Jacobian of the dependent variable ``ys`` versus the independent
+    variable ``xs``.
```
Comment (cxxly) [inline python/paddle/autograd/autodiff.py]: Computes the Hessian of ...

File: python/paddle/autograd/autodiff.py [hunk 6]
```diff
@@ -0,0 +1,748 @@
...
+    After the ``Hessian`` object is created, the complete calculation process does not
+    occur, but a partial lazy evaluation method is used for calculation. It can be
+    multi-dimensionally indexed to obtain the entire Hessian matrix or sub-matrix. At
+    this time, the actual Evaluates the computation and returns the result. At the same
+    time, in the actual evaluation process, the calculated sub-matrix will be cached to
+    avoid repeated calculations in the subsequent indexing process.
+
+    Args:
+
+        ys (paddle.Tensor): Output or list of outputs derived from xs.
```
Comment (cxxly) [inline python/paddle/autograd/autodiff.py]: ys should only a Tensor which contain one element

File: python/paddle/autograd/autodiff.py [hunk 7]
```diff
@@ -0,0 +1,748 @@
...
+        xs (Union[paddle.Tensor, Tuple[paddle.Tensor, ...]]): Input or list of inputs.
+        batch_axis (Optional[int], optional): Index of batch axis. Defaults to None.
+
+    Returns:
+
+        Union[List[List[Hessian]], List[Hessian], Hessian]: Hessian(s) of ys
+            deriveted from xs.
```
Comment (cxxly) [inline python/paddle/autograd/autodiff.py]: Tuple[Tuple...]

File: python/paddle/autograd/autodiff.py (hunk 1 above)
Comment (HydrogenSulfate) [reply python/paddle/autograd/autodiff.py]: Done

File: python/paddle/autograd/autodiff.py (hunk 2 above)
Comment (HydrogenSulfate) [reply python/paddle/autograd/autodiff.py]: Done

File: python/paddle/autograd/autodiff.py (hunk 4 above)
Comment (HydrogenSulfate) [reply python/paddle/autograd/autodiff.py]: Done

File: python/paddle/autograd/autodiff.py (hunk 6 above)
Comment (HydrogenSulfate) [reply python/paddle/autograd/autodiff.py]: Done

File: python/paddle/autograd/autodiff.py (hunk 7 above)
Comment (HydrogenSulfate) [reply python/paddle/autograd/autodiff.py]: Done

File: python/paddle/autograd/autodiff.py (hunk 3 above)
Comment (HydrogenSulfate) [reply python/paddle/autograd/autodiff.py]: Done

**Review by cxxly — DISMISSED** [review]
LGTM with TODO:

- add chinese document

**Review by cxxly — DISMISSED** [review]
LGTM

**Pull request closed without merging**